README.md
# jenkins-to-github-notify
Notify build status using a Jenkins Webhook


CHANGELOG.rst
=========
CHANGELOG
=========

0.1.3
=====

* Started accepting absolute job URLs.

0.1.2
=====

* Fixed entry-point parameters in Dockerfile.

0.1.1
=====

* Fixed listening port in Dockerfile.


0.1.0
=====

* First version.


tests/test_jenkins_to_github_notify.py
import json
from contextlib import contextmanager
from contextlib import nullcontext
from pathlib import Path

import github3
import jenkins
import pytest
import requests
from jenkins_to_github_notify.app import handle_jenkins_notification
from jenkins_to_github_notify.notify import BuildStatus
from jenkins_to_github_notify.notify import check_configuration
from jenkins_to_github_notify.notify import compute_job_alias
from jenkins_to_github_notify.notify import fetch_build_info
from jenkins_to_github_notify.notify import fetch_repo_infos_based_on_job_xml_config
from jenkins_to_github_notify.notify import parse_repos_and_branches
from jenkins_to_github_notify.notify import parse_slug
from jenkins_to_github_notify.notify import post_status_to_github
from jenkins_to_github_notify.notify import RepoBuildInfo
from jenkins_to_github_notify.notify import validate_event
from jenkins_to_github_notify.notify import validate_secret
from pytest_mock import MockerFixture


@pytest.fixture
def fake_config() -> dict[str, str]:
    return {
        "JENKINS_URL": "http://FAKE_JENKINS_URL",
        "JENKINS_USERNAME": "FAKE_JENKINS_USERNAME",
        "JENKINS_PASSWORD": "FAKE_JENKINS_PASSWORD",
        "JENKINS_SECRET": "FAKE_JENKINS_SECRET",
        "GH_TOKEN": "FAKE_GH_TOKEN",
    }


def test_check_configuration() -> None:
    config = {
        "JENKINS_URL": "",
        "JENKINS_USERNAME": "",
        "JENKINS_PASSWORD": "",
        "JENKINS_SECRET": "",
    }
    with pytest.raises(RuntimeError):
        check_configuration(config)

    config["GH_TOKEN"] = ""
    check_configuration(config)


def test_validate_secret() -> None:
    with pytest.raises(RuntimeError):
        validate_secret(dict(JENKINS_SECRET="secret1"), "secret2")

    validate_secret(dict(JENKINS_SECRET="secret2"), "secret2")


@pytest.mark.parametrize(
    "json_name, expected_status",
    [
        ("jenkins-response-1.json", BuildStatus.Success),
        ("jenkins-response-2.json", BuildStatus.Failure),
        ("jenkins-response-3.json", BuildStatus.Pending),
    ],
)
def test_fetch_build_status(
    fake_config: dict[str, str],
    datadir: Path,
    json_name: str,
    expected_status: BuildStatus,
    mocker: MockerFixture,
) -> None:
    job_name = "alfasim-fb-ASIM-4742-add-gui-support-esp-catolog-app-newlinux"
    build_number = 6
    with mocking_jenkins_build_info(
        mocker, datadir / json_name, fake_config, job_name, build_number
    ):
        result = fetch_build_info(fake_config, job_name, build_number)
        branch_name = "fb-ASIM-4742-add-gui-support-esp-catolog"
        assert list(result.repo_infos) == [
            RepoBuildInfo(
                slug="ESSS/alfasim",
                branch_name=branch_name,
                commit="183c3b5d60eb015704eb081d600b79e2c261f3a3",
            ),
            RepoBuildInfo(
                slug="ESSS/qmxgraph",
                branch_name=branch_name,
                commit="3a93b466b35fa703897d0d35fe93f12ea027da90",
            ),
            RepoBuildInfo(
                slug="ESSS/hookman",
                branch_name=branch_name,
                commit="4a7af78b5dc6d1bdd820ccea9c12beb07a113d13",
            ),
            RepoBuildInfo(
                slug="ESSS/alfasim-sdk",
                branch_name=branch_name,
                commit="de93939e21c4c942e2d0ad00d017c1df15b8f3ab",
            ),
        ]
        assert result.status is expected_status


def test_fetch_build_status_no_github_repos(
    fake_config: dict[str, str], datadir: Path, mocker: MockerFixture
) -> None:
    with mocking_jenkins_build_info(
        mocker,
        datadir / "jenkins-response-4.json",
        fake_config,
        "rocky20-fb-ROCKY-15386-contact-api-functors-esss-benchmark",
        6,
    ):
        result = fetch_build_info(
            fake_config, "rocky20-fb-ROCKY-15386-contact-api-functors-esss-benchmark", 6
        )
        assert list(result.repo_infos) == []
        assert result.status is BuildStatus.Failure


def test_fetch_build_status_no_build_info(
    fake_config: dict[str, str],
    datadir: Path,
    mocker: MockerFixture,
) -> None:
    """
    Fallback to ask GitHub for branches/commits in case the job build info from Jenkins
    does not contain that information.
    """
    job_name = "test-code-cov-fb-EDEN-2506-github-notification-newlinux"
    build_number = 16
    mock_login = mocker.patch.object(github3, "login")
    mock_repository = mock_login.return_value.repository
    mock_branch = mock_repository.return_value.branch
    mock_branch.return_value.commit.sha = "deadbeef"

    with mocking_jenkins_build_info(
        mocker, datadir / "jenkins-response-no-build-data.json", fake_config, job_name, build_number
    ) as jenkins_mock:
        job_xml_config = datadir.joinpath("job-multiple-repos.xml").read_text("UTF-8")
        jenkins_mock.return_value.get_job_config.return_value = job_xml_config
        result = fetch_build_info(fake_config, job_name, build_number)
        assert list(result.repo_infos) == [
            RepoBuildInfo(
                slug="ESSS/test-code-cov",
                branch_name="fb-EDEN-2506-github-notification",
                commit="deadbeef",
            )
        ]
        assert result.status is BuildStatus.Pending


@pytest.mark.parametrize(
    "input_url, expected_url",
    [
        ("ssh://git@github.com/ESSS/qmxgraph.git", "ESSS/qmxgraph"),
        ("git@github.com:ESSS/alfasim-sdk.git", "ESSS/alfasim-sdk"),
        ("https://github.com/ESSS/alfasim-sdk", "ESSS/alfasim-sdk"),
        ("git@github.com:ESSS/alfasim_sdk.git", "ESSS/alfasim_sdk"),
        ("https://github.com/ESSS/alfasim_sdk", "ESSS/alfasim_sdk"),
        ("git@otherserver.com:ESSS/alfasim.git", None),
        ("https://otherserver.com/ESSS/alfasim", None),
    ],
)
def test_parse_slug(input_url: str, expected_url: str | None) -> None:
    assert parse_slug(input_url) == expected_url


def test_validate_event() -> None:
    validate_event("jenkins.job.started")
    validate_event("jenkins.job.completed")


@pytest.mark.parametrize(
    "status_code, expected_outcome",
    [
        (201, nullcontext()),
        (301, pytest.raises(RuntimeError, match="ERROR 301")),
    ],
)
def test_post_status_to_github(
    mocker: MockerFixture, fake_config: dict[str, str], status_code: int, expected_outcome
) -> None:
    post_mock = mocker.patch.object(requests, "post")
    post_mock.return_value.status_code = status_code
    with expected_outcome:
        post_status_to_github(
            config=fake_config,
            slug="ESSS/test-code-cov",
            commit="80fd371bb50211b938afa7fa7c04f7b0f5fefecb",
            branch_name="fb-EDEN-2506-github-notification",
            job_name="test-code-cov-fb-EDEN-2506-github-notification-newlinux",
            job_url="job/test-code-cov-fb-EDEN-2506-github-notification-newlinux/8",
            build_number=8,
            status=BuildStatus.Success,
        )
    url = f"https://api.github.com/repos/ESSS/test-code-cov/statuses/80fd371bb50211b938afa7fa7c04f7b0f5fefecb"
    job_alias = "test-code-cov/newlinux"
    description = f"build #8 success"
    json_data = {
        "state": "success",
        "target_url": "http://FAKE_JENKINS_URL/job/test-code-cov-fb-EDEN-2506-github-notification-newlinux/8",
        "description": description,
        "context": f"{job_alias} job",
    }
    headers = {
        "Accept": "application/vnd.github+jso",
        "Authorization": f"token {fake_config['GH_TOKEN']}",
    }
    assert post_mock.call_args == mocker.call(url=url, json=json_data, headers=headers)


@pytest.mark.parametrize(
    "job_name, branch_name, expected_alias",
    [
        (
            "eden-fb-EDEN-2505-win64",
            "fb-EDEN-2505",
            "eden/win64",
        ),
        (
            "eden-fb-EDEN-2505-win64",
            "origin/fb-EDEN-2505",
            "eden/win64",
        ),
        ("eden-fb-EDEN-2505", "fb-EDEN-2505", "eden"),
    ],
)
def test_compute_job_alias(job_name: str, branch_name: str, expected_alias: str) -> None:
    assert compute_job_alias(job_name=job_name, branch_name=branch_name) == expected_alias


def test_handle_jenkins_notification(
    mocker: MockerFixture, fake_config: dict[str, str], datadir: Path
) -> None:
    """Integration test for the main endpoint."""
    import jenkins_to_github_notify.app

    job_name = "alfasim-fb-ASIM-4742-add-gui-support-esp-catolog-app-newlinux"
    build_number = 6

    with mocking_jenkins_build_info(
        mocker, datadir / "jenkins-response-1.json", fake_config, job_name, 6
    ):
        post_mock = mocker.patch.object(requests, "post")
        post_mock.return_value.status_code = 200

        mocker.patch.dict(jenkins_to_github_notify.app.config, values=fake_config, clear=True)

        handle_jenkins_notification(
            secret=fake_config["JENKINS_SECRET"],
            event="jenkins.job.started",
            job_name=job_name,
            build_number=build_number,
            url=f"job/{job_name}/{build_number}",
        )

    headers = {"Accept": "application/vnd.github+jso", "Authorization": "token FAKE_GH_TOKEN"}
    json = {
        "state": "success",
        "target_url": "http://FAKE_JENKINS_URL/job/alfasim-fb-ASIM-4742-add-gui-support-esp-catolog-app-newlinux/6",
        "description": "build #6 success",
        "context": "alfasim/app-newlinux job",
    }
    url_prefix = "https://api.github.com/repos/ESSS"
    assert post_mock.call_args_list == [
        mocker.call(
            url=f"{url_prefix}/alfasim/statuses/183c3b5d60eb015704eb081d600b79e2c261f3a3",
            json=json,
            headers=headers,
        ),
        mocker.call(
            url=f"{url_prefix}/qmxgraph/statuses/3a93b466b35fa703897d0d35fe93f12ea027da90",
            json=json,
            headers=headers,
        ),
        mocker.call(
            url=f"{url_prefix}/hookman/statuses/4a7af78b5dc6d1bdd820ccea9c12beb07a113d13",
            json=json,
            headers=headers,
        ),
        mocker.call(
            url=f"{url_prefix}/alfasim-sdk/statuses/de93939e21c4c942e2d0ad00d017c1df15b8f3ab",
            json=json,
            headers=headers,
        ),
    ]


class TestParseReposAndBranches:
    def test_parse_repos_and_branches_multiple(self, datadir: Path) -> None:
        job_xml_config = datadir.joinpath("job-multiple-repos.xml").read_text("UTF-8")
        result = parse_repos_and_branches(job_xml_config)
        assert result == [
            ("git@github.com:ESSS/test-code-cov.git", "fb-EDEN-2506-github-notification"),
            (
                "ssh://git@eden.fln.esss.com.br:7999/esss/eden.git",
                "fb-EDEN-2506-github-notification",
            ),
        ]

    def test_parse_repos_and_branches_single(self, datadir: Path) -> None:
        job_xml_config = datadir.joinpath("job-single-repo.xml").read_text("UTF-8")
        result = parse_repos_and_branches(job_xml_config)
        assert result == [
            ("ssh://git@eden.fln.esss.com.br:7999/esss/eden.git", "master"),
        ]


class TestFetchRepoInfosBasedOnJobXMLConfig:
    def test_fetch_repo_infos_based_on_job_xml_config(
        self, fake_config: dict[str, str], mocker: MockerFixture, datadir: Path
    ) -> None:
        mock_login = mocker.patch.object(github3, "login")
        mock_repository = mock_login.return_value.repository
        mock_branch = mock_repository.return_value.branch
        mock_branch.return_value.commit.sha = "deadbeef"

        job_xml_config = datadir.joinpath("job-multiple-repos.xml").read_text("UTF-8")

        result = fetch_repo_infos_based_on_job_xml_config(fake_config, job_xml_config)
        assert result == [
            RepoBuildInfo(
                slug="ESSS/test-code-cov",
                branch_name="fb-EDEN-2506-github-notification",
                commit="deadbeef",
            )
        ]
        assert mock_login.call_args == mocker.call(token=fake_config["GH_TOKEN"])
        assert mock_repository.call_args == mocker.call("ESSS", "test-code-cov")
        assert mock_branch.call_args == mocker.call("fb-EDEN-2506-github-notification")

    def test_fetch_repo_infos_based_on_job_xml_config_no_github_repo(
        self, fake_config: dict[str, str], mocker: MockerFixture, datadir: Path
    ) -> None:
        mock_login = mocker.patch.object(github3, "login")
        job_xml_config = datadir.joinpath("job-single-repo.xml").read_text("UTF-8")
        result = fetch_repo_infos_based_on_job_xml_config(fake_config, job_xml_config)
        assert result == []
        assert mock_login.call_args is None


@contextmanager
def mocking_jenkins_build_info(
    mocker: MockerFixture, response_json: Path, config: dict[str, str], job_name: str, number: int
):
    jenkins_mock = mocker.patch.object(jenkins, "Jenkins")
    response_data = json.loads(response_json.read_text("UTF-8"))
    jenkins_mock().get_build_info.return_value = response_data

    yield jenkins_mock

    assert jenkins_mock.call_args == mocker.call(
        config["JENKINS_URL"],
        username=config["JENKINS_USERNAME"],
        password=config["JENKINS_PASSWORD"],
    )
    assert jenkins_mock().get_build_info.call_args == mocker.call(job_name, number=number)


src/jenkins_to_github_notify/app.py
"""
FastAPI app entry point.

Here we declare all the FastAPI entry points, which should do minimal work as possible,
delegating to other modules the actual work.
"""
import logging

from dotenv import dotenv_values
from fastapi import FastAPI
from jenkins_to_github_notify.notify import check_configuration
from jenkins_to_github_notify.notify import fetch_build_info
from jenkins_to_github_notify.notify import post_status_to_github
from jenkins_to_github_notify.notify import validate_event
from jenkins_to_github_notify.notify import validate_secret

app = FastAPI()
config: dict[str, str] = {}
logger = logging.getLogger("app")


@app.on_event("startup")
async def startup_event() -> None:
    """Fail early during startup in case any variable is missing."""
    for key, value in dotenv_values(".env").items():
        assert isinstance(value, str), f"Unexpected type in config value: {value!r} {type(value)}"
        config[key] = value
    check_configuration(config)
    logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(levelname)s]  %(message)-8s")
    logger.setLevel(logging.INFO)
    logger.info("Configuration validated.")


@app.get("/jobs/notify")
def handle_jenkins_notification(
    secret: str,
    event: str,
    job_name: str,
    build_number: int,
    url: str,
) -> None:
    """
    End-point that is hit by Jenkins whenever an job starts or completes.

    It gathers all GitHub repositories that are checked out by the job and
    their hashes, and posts a status to each of their commits on GitHub.
    """
    logger.info("New job received")
    logger.info(f"  event: {event}")
    logger.info(f"  job_name: {job_name}")
    logger.info(f"  build_number: {build_number}")
    logger.info(f"  url: {url}")
    validate_secret(config, secret)
    validate_event(event)
    logger.info(f"  fetching build_info...")
    build_info = fetch_build_info(config, job_name, build_number)
    logger.info(f"  build_info: {build_info}")
    if build_info.repo_infos:
        logger.info(f"  posting status {build_info.status.value}")
    for repo_info in build_info.repo_infos:
        logger.info(f"  * {repo_info.slug} {repo_info.commit}")
        post_status_to_github(
            config=config,
            slug=repo_info.slug,
            commit=repo_info.commit,
            branch_name=repo_info.branch_name,
            job_name=job_name,
            build_number=build_number,
            status=build_info.status,
            job_url=url,
        )


src/jenkins_to_github_notify/notify.py
import re
import secrets
from enum import Enum
from io import StringIO
from typing import Any
from typing import Sequence
from urllib.parse import urljoin
from xml.etree.ElementTree import parse

import attr
import github3
import jenkins
import requests


def check_configuration(config: dict[str, str]) -> None:
    """
    Check that the given configuration dict contains all required variables.
    """

    def check_value(name: str) -> None:
        if name not in config:
            raise RuntimeError(f"Missing configuration variable: {name}.\nCheck your .env file.")

    check_value("JENKINS_URL")
    check_value("JENKINS_USERNAME")
    check_value("JENKINS_PASSWORD")
    check_value("JENKINS_SECRET")
    check_value("GH_TOKEN")


def validate_secret(config: dict[str, str], secret: str) -> None:
    """Ensures the secret provided by a Jenkins webhook matches what we expect in the configuration."""
    if not secrets.compare_digest(config["JENKINS_SECRET"], secret):
        raise RuntimeError("Invalid secret.")


def validate_event(event: str) -> None:
    """Ensures the event posted by the Jenkins webhook is one we expect."""
    if event not in ("jenkins.job.started", "jenkins.job.completed"):
        raise RuntimeError(f"Invalid event received: {event}")


class BuildStatus(Enum):
    """
    Status of a job.

    While Jenkins jobs have other possible status, for the purposes of this
    service we are only interested if it has failed or not after it has completed.
    """

    Pending = "pending"
    Success = "success"
    Failure = "failure"


@attr.s(auto_attribs=True, frozen=True, kw_only=True)
class RepoBuildInfo:
    """
    Information about a repository that is part of job run.
    """

    slug: str
    branch_name: str
    commit: str


@attr.s(auto_attribs=True, frozen=True)
class JobBuildInfo:
    """
    Information about a Jenkins job.
    """

    repo_infos: Sequence[RepoBuildInfo]
    status: BuildStatus


def fetch_build_info(config: dict[str, str], job_name: str, number: int) -> JobBuildInfo:
    """Fetches a JobBuildInfo structure containing GitHub repositories and commits for the given job."""
    server = jenkins.Jenkins(
        config["JENKINS_URL"],
        username=config["JENKINS_USERNAME"],
        password=config["JENKINS_PASSWORD"],
    )
    response_data = server.get_build_info(job_name, number=number)

    try:
        repo_infos = _extract_repo_infos_from_response(response_data)
    except NoRepositoriesFoundError:
        # Build info response does not contain any repositories/commits (happens when
        # starting a job), so we need to parse the repositories/branches configuration
        # from the job configuration and query GitHub.
        job_xml_config = server.get_job_config(job_name)
        repo_infos = fetch_repo_infos_based_on_job_xml_config(config, job_xml_config)

    job_result = response_data.get("result", "")
    if job_result is None:
        status = BuildStatus.Pending
    elif job_result == "SUCCESS":
        status = BuildStatus.Success
    else:
        status = BuildStatus.Failure
    return JobBuildInfo(repo_infos, status)


def _extract_repo_infos_from_response(response_data: dict[str, Any]) -> Sequence[RepoBuildInfo]:
    """
    Extract a sequence of RepoBuildInfo from the the  response of a get_build_info() call
    for a jenkins job.

    This returns only build information for GitHub repositories, any other servers are ignored.

    Raises NoRepositoriesFoundError if there is no repository at all definition in the response (GitHub or not).
    """
    repo_infos = []
    found_any_repos_at_all = False
    for action in response_data.get("actions", []):
        if action.get("_class", "") != "hudson.plugins.git.util.BuildData":
            continue
        remote_urls = action.get("remoteUrls", [])
        if not remote_urls:
            continue
        found_any_repos_at_all = True
        for remote_url in remote_urls:
            if slug := parse_slug(remote_url):
                branch_data = action["lastBuiltRevision"]["branch"][0]
                branch_name = branch_data["name"].removeprefix("origin/")
                commit = branch_data["SHA1"]
                repo_infos.append(RepoBuildInfo(slug=slug, commit=commit, branch_name=branch_name))

    if found_any_repos_at_all:
        return repo_infos
    else:
        raise NoRepositoriesFoundError


def fetch_repo_infos_based_on_job_xml_config(
    config: dict[str, str], job_xml_config: str
) -> Sequence[RepoBuildInfo]:
    """
    Parse the given Jenkins job configuration and extract the GitHub repositories and branches from it,
    using then the GitHub API to query the latest commit for that branch.
    """
    repos_and_branches = parse_repos_and_branches(job_xml_config)
    slugs_and_branches = [
        (parse_slug(url), branch) for url, branch in repos_and_branches if parse_slug(url)
    ]
    if not slugs_and_branches:
        return []

    result = []
    gh = github3.login(token=config["GH_TOKEN"])
    for slug, branch in slugs_and_branches:
        assert slug is not None
        owner, name = slug.split("/")
        repo = gh.repository(owner, name)
        sha = repo.branch(branch).commit.sha
        result.append(RepoBuildInfo(slug=slug, branch_name=branch, commit=sha))

    return result


def parse_repos_and_branches(job_xml_config: str) -> Sequence[tuple[str, str]]:
    """
    Given the XML configuration of a job, parse it and return a sequence of
    (repo-url, branch-name).
    """
    et = parse(StringIO(job_xml_config))
    r = et.getroot()

    # Multiple repositories:
    git_roots = list(r.findall(".//hudson.plugins.git.GitSCM"))
    # Single repositories:
    git_roots += list(r.findall(".//scm[@class='hudson.plugins.git.GitSCM']"))

    result = []
    for git_root in git_roots:
        names = [x.text for x in git_root.findall(".//hudson.plugins.git.BranchSpec/name")]
        urls = [x.text for x in git_root.findall(".//hudson.plugins.git.UserRemoteConfig/url")]
        if names and urls:
            name = names[0]
            url = urls[0]
            if name and url:
                result.append((url, name))
    return result


def parse_slug(url: str) -> str | None:
    """
    Parses the GitHub slug from the given HTTPS or SSH URLs (a slug is a string in the form "owner/repo").
    If it is not a valid GitHub address returns None.
    """
    if m := re.match(r"(?:ssh://)?git@github.com[:/]([\w_-]+)/([\w_-]+)\.git", url):
        return f"{m.group(1)}/{m.group(2)}"
    elif m := re.match(r"https://github.com/([\w_-]+)/([\w_-]+)", url):
        return f"{m.group(1)}/{m.group(2)}"
    else:
        return None


def post_status_to_github(
    *,
    config: dict[str, str],
    slug: str,
    commit: str,
    branch_name: str,
    job_name: str,
    job_url: str,
    build_number: int,
    status: BuildStatus,
) -> None:
    url = f"https://api.github.com/repos/{slug}/statuses/{commit}"
    job_alias = compute_job_alias(job_name=job_name, branch_name=branch_name)
    description = f"build #{build_number} {status.value}"
    json_data = {
        "state": status.value,
        "target_url": urljoin(config["JENKINS_URL"], job_url),
        "description": description,
        "context": f"{job_alias} job",
    }
    headers = {
        "Accept": "application/vnd.github+jso",
        "Authorization": f"token {config['GH_TOKEN']}",
    }
    response = requests.post(url=url, json=json_data, headers=headers)
    if not (200 <= response.status_code < 300):
        raise RuntimeError(f"ERROR {response.status_code} ({response.json()})")


def compute_job_alias(*, job_name: str, branch_name: str) -> str:
    """
    Given a full job name and branch name, returns a more compact version
    to display in the PR.

    For example:

        >>> compute_job_alias(job_name="alfasim-fb-EDEN-2505-app-win64", branch_name="fb-EDEN-2505")
        "alfasim/app-win64"
    """
    origin_prefix = "origin/"
    if branch_name.startswith(origin_prefix):
        branch_name = branch_name[len(origin_prefix) :]
    # Remove branch name from the job name:
    # "alfasim-fb-EDEN-2505-app-win64" -> "alfasim--app-win64"
    alias = job_name.replace(branch_name, "")
    # Change "--" from removing the branch to "/" for a nice visual separation:
    # "alfasim--app-win64" -> "alfasim/app-win64"
    alias = alias.replace("--", "/")
    # Remove any trailing "-" in case the branch is the last part:
    # "test-repo-fb-EDEN-2505" -> "test-repo-" -> "test-repo"
    alias = alias.rstrip("-")
    return alias


class NoRepositoriesFoundError(Exception):
    """
    Raised by _extract_repo_infos_from_response if not repositories are found in
    the response.
    """
